lib/web/proxy_settings.go

```diff
@@ -0,0 +1,161 @@
+		},
+	}
+
+	p.setProxyPublicAddressesSettings(&proxySettings)
+
+	if !p.ServiceConfig.Proxy.MySQLAddr.IsEmpty() {
+		proxySettings.DB.MySQLListenAddr = p.ServiceConfig.Proxy.MySQLAddr.String()
+	}
+
+	if !p.ServiceConfig.Proxy.PostgresAddr.IsEmpty() {
+		proxySettings.DB.PostgresListenAddr = p.ServiceConfig.Proxy.PostgresAddr.String()
+	}
+
+	if !p.ServiceConfig.Proxy.MongoAddr.IsEmpty() {
+		proxySettings.DB.MongoListenAddr = p.ServiceConfig.Proxy.MongoAddr.String()
+	}
+
+	if p.ServiceConfig.Proxy.Kube.Enabled {
+		proxySettings.Kube.ListenAddr = p.ServiceConfig.Proxy.Kube.ListenAddr.String()
+	}
+	return &proxySettings
+}
+
+// buildProxySettingsV2 builds the v2 proxy settings where teleport proxies can start only on a single listener.
+func (p *ProxySettings) buildProxySettingsV2(proxyListenerMode types.ProxyListenerMode, sshDialTimeout time.Duration) *webclient.ProxySettings {
+	multiplexAddr := p.ServiceConfig.Proxy.WebAddr.String()
+	settings := p.buildProxySettings(proxyListenerMode, sshDialTimeout)
+	if proxyListenerMode == types.ProxyListenerMode_Multiplex {
+		settings.SSH.ListenAddr = multiplexAddr
+		settings.SSH.TunnelListenAddr = multiplexAddr
+		settings.SSH.WebListenAddr = multiplexAddr
+		settings.Kube.ListenAddr = multiplexAddr
+		settings.DB.MySQLListenAddr = multiplexAddr
+		settings.DB.PostgresListenAddr = multiplexAddr
+	}
+	return settings
+}
+
+func (p *ProxySettings) setProxyPublicAddressesSettings(settings *webclient.ProxySettings) {
+	if len(p.ServiceConfig.Proxy.PublicAddrs) > 0 {
+		settings.SSH.PublicAddr = p.ServiceConfig.Proxy.PublicAddrs[0].String()
+	}
+	if len(p.ServiceConfig.Proxy.SSHPublicAddrs) > 0 {
+		settings.SSH.SSHPublicAddr = p.ServiceConfig.Proxy.SSHPublicAddrs[0].String()
+	}
+	if len(p.ServiceConfig.Proxy.TunnelPublicAddrs) > 0 {
+		settings.SSH.TunnelPublicAddr = p.ServiceConfig.Proxy.TunnelPublicAddrs[0].String()
+	}
+	if len(p.ServiceConfig.Proxy.Kube.PublicAddrs) > 0 {
+		settings.Kube.PublicAddr = p.ServiceConfig.Proxy.Kube.PublicAddrs[0].String()
+	}
+	if len(p.ServiceConfig.Proxy.MySQLPublicAddrs) > 0 {
+		settings.DB.MySQLPublicAddr = p.ServiceConfig.Proxy.MySQLPublicAddrs[0].String()
+	}
+	if len(p.ServiceConfig.Proxy.MongoPublicAddrs) > 0 {
+		settings.DB.MongoPublicAddr = p.ServiceConfig.Proxy.MongoPublicAddrs[0].String()
+	}
+	settings.DB.PostgresPublicAddr = p.getPostgresPublicAddr()
+}
+
+// getPostgresPublicAddr returns the proxy PostgresPublicAddrs based on whether the Postgres proxy service
+// was configured on separate listener. For backward compatibility if PostgresPublicAddrs was not provided.
+// Proxy will reuse the PostgresPublicAddrs field to propagate postgres service address to legacy tsh clients.
+func (p *ProxySettings) getPostgresPublicAddr() string {
+	if len(p.ServiceConfig.Proxy.PostgresPublicAddrs) > 0 {
+		return p.ServiceConfig.Proxy.PostgresPublicAddrs[0].String()
+	}
+
+	if p.ServiceConfig.Proxy.PostgresAddr.IsEmpty() {
+		return ""
+	}
+
+	// DELETE IN 9.0.0
```

Comment:
Out of scope for this PR, but maybe this is a good reminder to get someone assigned to remove this.